Essentials/src/main/java/com/earth2me/essentials/Kit.java

```diff
@@ -214,57 +213,34 @@ public boolean expandItems(final User user, final List<String> items) throws Exc
-                        inventory.setLeggings(stack);
-                        continue;
-                    } else if (MaterialUtil.isBoots(material) && isEmptyStack(inventory.getBoots())) {
-                        inventory.setBoots(stack);
-                        continue;
-                    }
-                }
-
                 itemList.add(stack);
             }
 
-            final Map<Integer, ItemStack> overfilled;
-            final boolean allowOversizedStacks = user.isAuthorized("essentials.oversizedstacks");
+            final int maxStackSize = user.isAuthorized("essentials.oversizedstacks") ? ess.getSettings().getOversizedStackSize() : 0;
             final boolean isDropItemsIfFull = ess.getSettings().isDropItemsIfFull();
-            if (isDropItemsIfFull) {
-                if (allowOversizedStacks) {
-                    overfilled = InventoryWorkaround.addOversizedItems(user.getBase().getInventory(), ess.getSettings().getOversizedStackSize(), itemList.toArray(new ItemStack[0]));
-                } else {
-                    overfilled = InventoryWorkaround.addItems(user.getBase().getInventory(), itemList.toArray(new ItemStack[0]));
-                }
-                for (final ItemStack itemStack : overfilled.values()) {
-                    int spillAmount = itemStack.getAmount();
-                    if (!allowOversizedStacks) {
-                        itemStack.setAmount(Math.min(spillAmount, itemStack.getMaxStackSize()));
-                    }
-                    while (spillAmount > 0) {
-                        user.getWorld().dropItemNaturally(user.getLocation(), itemStack);
-                        spillAmount -= itemStack.getAmount();
-                    }
-                    spew = true;
-                }
-            } else {
-                if (allowOversizedStacks) {
-                    overfilled = InventoryWorkaround.addAllOversizedItems(user.getBase().getInventory(), ess.getSettings().getOversizedStackSize(), itemList.toArray(new ItemStack[0]));
-                } else {
-                    overfilled = InventoryWorkaround.addAllItems(user.getBase().getInventory(), itemList.toArray(new ItemStack[0]));
+
+            final ItemStack[] itemArray = itemList.toArray(new ItemStack[0]);
+            if (!isDropItemsIfFull && !Inventories.hasSpace(user.getBase(), maxStackSize, autoEquip, itemArray)) {
+                user.sendMessage(tl("kitInvFullNoDrop"));
+                return false;
+            }
+
+            final Map<Integer, ItemStack> leftover = Inventories.addItem(user.getBase(), maxStackSize, autoEquip, itemArray);
+            if (!isDropItemsIfFull && !leftover.isEmpty()) {
+                // Inventories#hasSpace should prevent this state from EVER being reached; If it does, something has gone terribly wrong, and we should just give up and hope people report it :(
+                throw new IllegalStateException(leftover.toString());
```

Comment:
Can we make a more user-friendly error message for this instead of just dumping the items